pkg/controller/namespace_helper.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2023 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+
+	"github.com/tigera/operator/pkg/common"
+	"github.com/tigera/operator/pkg/controller/utils"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func NewSingleTenantNamespaceHelper(ns string) NamespaceHelper {
+	return &namespacer{
+		multiTenant:           false,
+		singleTenantNamespace: ns,
+	}
+}
+
+func NewNamespaceHelper(mt bool, singleTenantNS, multiTenantNS string) NamespaceHelper {
+	return &namespacer{
+		multiTenant:           mt,
+		singleTenantNamespace: singleTenantNS,
+		multiTenantNamespace:  multiTenantNS,
+	}
+}
+
+type NamespaceHelper interface {
+	// InstallNamespace returns the namespace that components will be installed into.
+	// for single-tenant clusters, this is tigera-manager. For multi-tenancy, this
+	// will be the tenant's namespace.
+	InstallNamespace() string
+
+	// TruthNamespace returns the namespace to use as the source of truth for storing data.
+	// For single-tenant installs, this is the tigera-operator namespace.
+	// For multi-tenant installs, this is tenant's namespace.
+	TruthNamespace() string
```

Comment:
I'm not a big fan of the name, but Im not sure if I know a better option.

To me the truth namespace is not so much about being the truth, but more about the namespace where users can provide input. So perhaps userInputNamespace, configurationNamespace or inputNamespace.

Perhaps the installNamespace could be called componentNamespace or managerNamespace(). The installation CR is "installing" in the calico-system namespace, making installNamespace all the more confusing.

Reply:
It's not really about whether the config was done by the user or not, it's about whether the config is the canonical representation or if it has been copied from the canonical representation.

We could call it the "SpecNamespace" to be more consistent with Kubernetes terms - e.g., spec vs. status. - but I think it's less clear than "TruthNamespace", which itself is less clear (and less verbose) than "SourceOfTruthNamespace"